### PR TITLE
[Errors] [WIP] Error codes CLI utility tool

### LINKED ISSUE #46: [Errors] Error codes utility script
**Current Behavior**

Many error codes are placeholders that need to be replaced. There is no static check that verifies that codes within a repo/component do not overlap.

**Desired Behavior**

Script to verify (no duplicates) and update error codes in source code. Also to extract error information into a file that can be used as basis for publishing all error codes on the meshery website. Can be run locally and as part of a CI workflow.

---
**Imlementation notes**

see https://github.com/layer5io/meshkit/pull/48

---
**Resources**

https://docs.google.com/document/d/1_zpEWBcC6ngOLen_E_nc-octiUdBbpmdwwIX6qzo70M/edit

**Alternatives / Additional Context**

### PR DESCRIPTION
**Pull request by mgfeller**
**Description**

This PR fixes #46

Tool to verify and update error codes in source code. Also to extract error details into a file that can be used as basis for publishing all error codes on the Meshery website (see https://github.com/layer5io/meshery/pull/2712). 

Intended to be run locally and as part of a CI workflow.

Build binary using `make errorutil`.

Includes also fixes for long-standing golangci-lint issues. (**reverted**, will be implemented in a separate PR)

**Notes for Reviewers**

This PR is not yet ready to be merged. 

Remaining work:
* some code cleanup
* improved tool and command documentation
* exporting error details from errors.New(...) 
* possibly running the tool on `meshkit` and committing the result and cleaned up error codes, with some detailed info added.
* changes based on review

**Implementation notes**

- errors names and codes (an int) are namespaced to components, i.e. need to be unique within a component, e.g. the istio adapter. the tool verifies it for code already, and will for error names.
- there are no predefined error code ranges for components - every component is free to use its own range, but it looks like the convention is to start at 1000
- codes carry no meaning, as e.g. http status codes do
- in the code, create string vars or consts with names starting with `Err[A-Z]`
- set the value to any string, like `"replace_me"` (no convention here), e.g. `ErrApplyManifestCode    = "test_code"`
- if the value is an int, e.g. `ErrGetName        = "1000"`  the tool will not replace it unless it is forced (command line flag), then all codes are renumbered (useful e.g. in `meshkit).`
- setting an error code to a call expression like this` ErrNoneDatabase = errors.NewDefault(ErrNoneDatabaseCode, "No Database selected") `is not allowed, the tool checks that.
- using `errors.NewDefault(...)` is deprecated, the tool checks that and emits a warning
- use `errors.New(...)` from `meshkit` to create actual errors, specify detailed descriptions, probable causes, remedies. the tool extracts this from the code. this is often done in a factory function, it is important that the error code variable is used here, not a literal:
```
func ErrNewTestTwo(err error) error {
    return errors.New(ErrTestCode2, errors.Fatal, []string{"line1"}, []string{"line1", "line2"}, []string{}, []string{"line1"}) 
 }
 ```
- by conventions, error codes and the factory functions live in files called `error.go`, but the tool checks all files (could be controlled by a flag).
- running the tool will create a couple of files, one of them designed to be used to generate the error reference on the meshery website.
- The tool requires a file called `component_info.json` in (the root of) the repo, with the following content (different for each repo, of course):
```
{
  "name": "meshkit",
  "type": "library",
  "next_error_code": 11010
}
```
- The tool updates `next_error_code`.

Go's ast library is used.

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -5,4 +5,7 @@ check-clean-cache:
 	golangci-lint cache clean
 
 test:
-	go test ./...
\ No newline at end of file
+	go test ./...
+
+errorutil:
```
Comment (kumarabd) [inline Makefile:10]: @mgfeller Can this be a `go generate <file>`?

File: cmd/errorutil/internal/config/app.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package config
+
+const (
+	App     = "errorutil"
+	LogFile = App + ".log"
```
Comment (kumarabd) [inline cmd/errorutil/internal/config/app.go:5]: Any reason for creating a log file instead of stdout?

File: Makefile (hunk 1 above)
Comment (mgfeller) [reply Makefile:10]: @kumarabd just had a look... no I don't think so.

File: cmd/errorutil/internal/config/app.go (hunk 2 above)
Comment (mgfeller) [reply cmd/errorutil/internal/config/app.go:5]: Yes. Old habit for CLI applications, log to file (also to have history over multiple invocations) and more selective output to console. Don't like getting all log on stdout when using a cli tool. Haven't made up my mind here yet, though, and not finished yet :smile:

**Comment by kumarabd** [comment]
@mgfeller Is it still `do-not-merge`?

**Comment by leecalcote** [comment]
Please say, "No". Please say, "No".

**Comment by mgfeller** [comment]
@kumarabd @leecalcote yes, it is still `do-not-merge`. Sorry for the delay, moving house is a long process and a lot of work. First lap almost completed at home - and here I assume I can finalize this work at the beginning of next week :smile: 

// @alonsolopezr

**Comment by mgfeller** [comment]
@kumarabd now it is ready to be reviewed and merged if approved  :smile:  
// @leecalcote

**Review by navendu-pottekkat — APPROVED** [review]
Looks good. Will look into using this in mesheryctl

**Comment by metonymic-smokey** [comment]
Looks good to me too!
@navendu-pottekkat using these could be part of the mesheryctl refactor?

**Comment by navendu-pottekkat** [comment]
> Looks good to me too!
> @navendu-pottekkat using these could be part of the mesheryctl refactor?

Yes, we can get started on this

**Pull request merged**